### ISSUE TITLE
[no-unnecessary-type-assertion] False positive with Readonly<T>

### ISSUE THREAD

**Issue by PSeitz** [typescript-eslint/typescript-eslint#1595]
```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-type-assertion": "warn"
  }
}
```

```TS
  export function upgradeV37Schema(input: Readonly<string|undefined>) {
    return input!;
  }
```

**Actual Result**
`This assertion is unnecessary since it does not change the type of the expression.eslint(@typescript-eslint/no-unnecessary-type-assertion)`

No issue when Readonly is removed.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.18.0` |
| `@typescript-eslint/parser`        | `2.18.0` |
| `TypeScript`                       | `3.7.3` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `10.15.3` |
| `npm`                              | `6.4.1` |

**Comment by bradzacher** [comment]
I am unable to reproduce this against master.

![image](https://user-images.githubusercontent.com/7462525/74362615-dd767480-4d7d-11ea-9642-8be1a8627b4f.png)

Are you using `strictNullChecks`?

**Comment by PSeitz** [comment]
I think the vscode plugin got me again, I had to restart vscode and the error dissappeared. Some screenshots to prove I'm not imagining things ;)

Any idea how this could happen?

<img width="640" alt="Screenshot 2020-02-12 at 20 16 11" src="https://user-images.githubusercontent.com/1109503/74369150-27d71000-4dd5-11ea-912d-11ae43df3f23.png">

<img width="697" alt="Screenshot 2020-02-12 at 20 16 46" src="https://user-images.githubusercontent.com/1109503/74369168-2efe1e00-4dd5-11ea-867a-c15fb576ea88.png">

**Comment by bradzacher** [comment]
There's probably a caching thing somewhere. There's many parts to the plugin system, and each provides some level of caching to try and be as quick as possible.

I would recommend that if you're ever in doubt to save the file and run the lint from the console. The lint should never apply caching, and should build state from fresh, so you can validate if the error actually exists.
